### ISSUE TITLE
Add the ability to manage Studio Flow definition JSON via Terraform

### ISSUE THREAD

**Issue by RJPearson94** [RJPearson94/terraform-provider-twilio#31]
## Is your feature request related to a problem

No, not related to a problem. 

Currently, you can manage a studio flow using the `twilio_studio_flow` resource, you can define the flow definition JSON inline or load it from an external file, the state transitions are predefined and interpolation can be used to substitute in certain values. 

This approach is fine but if you need to conditionally include transitions e.g. in certain environments, you may not want to include a specific widget/ resources i.e. send to Autopilot. This would mean you would need to manage separate configuration for each permutation of the flow which would bring additional maintenance overhead. 

The provider should support managing Studio flow definition and widgets as data sources as this would allow meta-arguments i.e. count, expressions, functions, etc. to be used to dynamically generate the definition JSON. This JSON can then be supplied as an input to the `twilio_studio_flow` resource.

## Describe the solution you'd like

There would be a new data source for the Studio Flow definition and each Studio Flow widget. 

The example below shows creating a studio flow with the `trigger` and `send to flex` widgets

```hcl
## Studio Flow Definition 

data "twilio_studio_flow_widget_send_to_flex" "send_to_flex" {
  name = "SendMessageToAgent"

  workflow = var.workflow_sid
  channel  = var.channel_sid
  attributes = jsonencode({
    "name" : "{{trigger.message.ChannelAttributes.from}}",
    "channelType" : "{{trigger.message.ChannelAttributes.channel_type}}",
    "channelSid" : "{{trigger.message.ChannelSid}}"
  })

  offset {
    x = 200
    y = 240
  }
}

data "twilio_studio_flow_widget_trigger" "trigger" {
  name = "Trigger"

  transitions {
    incoming_message = data.twilio_studio_flow_widget_send_to_flex.send_to_flex.name
  }

  offset {
    x = 200
    y = 0
  }
}

data "twilio_studio_flow_definition" "definition" {
  description   = "Bot flow for creating a Flex webchat task"
  initial_state = data.twilio_studio_flow_widget_trigger.trigger.name

  flags {
    allow_concurrent_calls = true
  }

  state {
    json = data.twilio_studio_flow_widget_send_to_flex.send_to_flex.json
  }

  state {
    json = data.twilio_studio_flow_widget_trigger.trigger.json
  }
}

## Studio Flow 

resource "twilio_studio_flow" "flow" {
  friendly_name = "With widgets"
  status        = "draft"
  definition    = data.twilio_studio_flow_definition.definition.json
  validate      = true
}

## Variables

variable "workflow_sid" {
  description = "Task Router Workflow sid"
  type        = string
}

variable "channel_sid" {
  description = "Task Router Channel sid"
  type        = string
}

```

The example below shows the same studio flow above with the `send to autopilot` widget being conditionally included as the first transition in the definition JSON

```hcl
## Studio Flow Definition 

locals {
  include_autopilot = var.autopilot_assistant_sid != null

  states = compact([
    data.twilio_studio_flow_widget_trigger.trigger.json,
    data.twilio_studio_flow_widget_send_to_flex.send_to_flex.json,
    join("", data.twilio_studio_flow_widget_send_to_autopilot.send_to_autopilot.*.json),
  ])

  message_transition = coalesce(
    join("", data.twilio_studio_flow_widget_send_to_autopilot.send_to_autopilot.*.name),
    data.twilio_studio_flow_widget_send_to_flex.send_to_flex.name
  )
}

data "twilio_studio_flow_widget_send_to_autopilot" "send_to_autopilot" {
  count = local.include_autopilot ? 1 : 0
  name  = "SendToAutopilot"

  offset {
    x = 200
    y = 240
  }

  transitions {
    failure = data.twilio_studio_flow_widget_send_to_flex.send_to_flex.name
    timeout = data.twilio_studio_flow_widget_send_to_flex.send_to_flex.name
  }

  autopilot_assistant_sid = var.autopilot_assistant_sid
  from                    = "{{flow.channel.address}}"
  body                    = "{{trigger.message.Body}}"
  timeout                 = 14400
}

data "twilio_studio_flow_widget_send_to_flex" "send_to_flex" {
  name = "SendMessageToAgent"

  workflow = var.workflow_sid
  channel  = var.channel_sid
  attributes = jsonencode({
    "name" : "{{trigger.message.ChannelAttributes.from}}",
    "channelType" : "{{trigger.message.ChannelAttributes.channel_type}}",
    "channelSid" : "{{trigger.message.ChannelSid}}"
  })

  offset {
    x = 270
    y = 540
  }
}

data "twilio_studio_flow_widget_trigger" "trigger" {
  name = "Trigger"

  transitions {
    incoming_message = local.message_transition
  }

  offset {
    x = 200
    y = 0
  }
}

data "twilio_studio_flow_definition" "definition" {
  description   = "Bot flow for creating a Flex webchat task"
  initial_state = data.twilio_studio_flow_widget_trigger.trigger.name

  flags {
    allow_concurrent_calls = true
  }

  dynamic "state" {
    for_each = local.states
    content {
      json = state.value
    }
  }
}

## Studio Flow 

resource "twilio_studio_flow" "flow" {
  friendly_name = "Conditional Widgets"
  status        = "draft"
  definition    = data.twilio_studio_flow_definition.definition.json
  validate      = true
}

## Variables

variable "workflow_sid" {
  description = "Task Router Workflow sid"
  type        = string
}

variable "channel_sid" {
  description = "Task Router Channel sid"
  type        = string
}

variable "autopilot_assistant_sid" {
  description = "autopilot assistant sid"
  type        = string
  default     = null
}
```

The solution should not enforce the use of the widget data sources i.e. if a user wants to specify the widget definition inline or in external files then this should be supported too.

**Comment by timdodsgon** [comment]
I work for a large insurance company in the uk, we use Twilio studio flow and have to manually combine subflows for different customers, this sound very interesting.

**Comment by RJPearson94** [comment]
Due to timelines, I am going to push this back to v0.11.0. This allows me to release a bug fix and some changes which improve the developer experience when working with the provider

**Comment by RJPearson94** [comment]
This has been released as part of v0.11.0 🎉 

There have been a few adjustments to argument names in the final implementation so please see the docs and examples for more information
